presto-main-base/src/main/java/com/facebook/presto/server/PluginManagerUtil.java

```diff
@@ -0,0 +1,274 @@
+        log.debug("Classpath for %s:", dir.getName());
+        List<URL> urls = new ArrayList<>();
+        for (File file : listFiles(dir)) {
+            log.debug("    %s", file);
+            urls.add(file.toURI().toURL());
+        }
+        return createClassLoader(urls, spiPackages, parent);
+    }
+
+    private static URLClassLoader buildClassLoaderFromCoordinates(
+            String coordinates,
+            ArtifactResolver resolver,
+            List<String> spiPackages,
+            ClassLoader parent)
+            throws Exception
+    {
+        Artifact rootArtifact = new DefaultArtifact(coordinates);
+        List<Artifact> artifacts = resolver.resolveArtifacts(rootArtifact);
+        return createClassLoader(artifacts, rootArtifact.toString(), spiPackages, parent);
+    }
+
+    private static URLClassLoader createClassLoader(
+            List<Artifact> artifacts,
+            String name,
+            List<String> spiPackages,
+            ClassLoader parent)
+            throws IOException
+    {
+        log.debug("Classpath for %s:", name);
+        List<URL> urls = new ArrayList<>();
+        for (Artifact artifact : sortedArtifacts(artifacts)) {
+            if (artifact.getFile() == null) {
+                throw new RuntimeException("Could not resolve artifact: " + artifact);
+            }
+            File file = artifact.getFile().getCanonicalFile();
+            log.debug("    %s", file);
+            urls.add(file.toURI().toURL());
+        }
+        return createClassLoader(urls, spiPackages, parent);
+    }
+
+    private static URLClassLoader createClassLoader(List<URL> urls, List<String> spiPackages, ClassLoader parent)
+    {
+        return new PluginClassLoader(urls, parent, spiPackages);
+    }
+
+    private static List<File> listFiles(File installedPluginsDir)
+    {
+        if (installedPluginsDir != null && installedPluginsDir.isDirectory()) {
+            File[] files = installedPluginsDir.listFiles();
+            if (files != null) {
+                Arrays.sort(files);
+                return ImmutableList.copyOf(files);
+            }
+        }
+        return ImmutableList.of();
+    }
+
+    private static List<Artifact> sortedArtifacts(List<Artifact> artifacts)
+    {
+        List<Artifact> list = new ArrayList<>(artifacts);
+        Collections.sort(list, Ordering.natural().nullsLast().onResultOf(Artifact::getFile));
+        return list;
+    }
+
+    private static void processPlugins(Artifact artifact, ClassLoader classLoader, String servicesFile, String className)
+            throws IOException
+    {
+        Set<String> plugins = discoverPlugins(artifact, classLoader, servicesFile, className);
+        if (!plugins.isEmpty()) {
+            writePluginServices(plugins, artifact.getFile(), servicesFile);
+        }
+    }
+
+    public interface PluginInstaller
+    {
+        void installPlugin(Plugin plugin);
+
+        void installCoordinatorPlugin(CoordinatorPlugin plugin);
```

Comment:
not sure if this is the right abstraction since `installCoordinatorPlugin` would have no bearing on the router. I like this idea though. Let's discuss offline